File: 24.swap-nodes-in-pairs.rs
/*
 * @lc app=leetcode id=24 lang=rust
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() || head.as_ref().unwrap().next.is_none() {
            return head;
        }
    
        let mut first = head;
        let mut  second = first.as_mut().unwrap().next.take();
        let third = second.as_mut().unwrap().next.take();
        let reset = Self::swap_pairs(third);

        first.as_mut()?.next = reset;
        second.as_mut()?.next = first;

        second
    }
}
// @lc code=end

File: 24.swap-nodes-in-pairs.go
package main

/*
 * @lc app=leetcode id=24 lang=golang
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	nextNode := head.Next
	head.Next = swapPairs(head.Next.Next)
	nextNode.Next = head

	return nextNode
}

// @lc code=end

type ListNode struct {
	Val  int
	Next *ListNode
}
